firebase-devservices/deployment/src/main/java/io/quarkiverse/googlecloudservices/firebase/deployment/testcontainers/FirebaseEmulatorContainer.java

```diff
@@ -0,0 +1,1382 @@
+
+            emulatorConfig.customFirebaseJson().ifPresentOrElse(
+                    this::includeCustomFirebaseJson,
+                    this::generateFirebaseJson);
+
+            this.dockerBuilder.add("firebase.json", FIREBASE_ROOT + "/firebase.json");
+        }
+
+        private void includeCustomFirebaseJson(Path customFilePath) {
+            this.result.withFileFromPath(
+                    "firebase.json",
+                    customFilePath);
+        }
+
+        private void includeFirestoreFiles() {
+            emulatorConfig.firebaseConfig().firestoreConfig.rulesFile.ifPresent(rulesFile -> {
+                this.dockerBuilder.add("firestore.rules", FIREBASE_ROOT + "/firestore.rules");
+                this.result.withFileFromPath("firestore.rules", rulesFile);
+            });
+
+            emulatorConfig.firebaseConfig().firestoreConfig.indexesFile.ifPresent(indexesFile -> {
+                this.dockerBuilder.add("firestore.indexes.json", FIREBASE_ROOT + "/firestore.indexes.json");
+                this.result.withFileFromPath("firestore.indexes.json", indexesFile);
+            });
+        }
+
+        private void includeStorageFiles() {
+            emulatorConfig.firebaseConfig().storageConfig.rulesFile.ifPresent(rulesFile -> {
+                this.dockerBuilder.add("storage.rules", FIREBASE_ROOT + "/storage.rules");
+                this.result.withFileFromPath("storage.rules", rulesFile);
+            });
+        }
+
+        private void generateFirebaseJson() {
+            var firebaseJsonBuilder = new FirebaseJsonBuilder(this.emulatorConfig);
+            String firebaseJson;
+            try {
+                firebaseJson = firebaseJsonBuilder.buildFirebaseConfig();
+            } catch (IOException e) {
+                throw new IllegalStateException("Failed to generate firebase.json file", e);
+            }
+
+            this.result.withFileFromString("firebase.json", firebaseJson);
+        }
+
+        private void setupDataImportExport() {
+            emulatorConfig.cliArguments().emulatorData().ifPresent(
+                    emulator -> this.dockerBuilder.volume(EMULATOR_DATA_PATH));
+        }
+
+        private void setupHosting() {
+            // Specify public directory if hosting is enabled
+            if (emulatorConfig.firebaseConfig().hostingConfig().hostingContentDir().isPresent()) {
+                this.dockerBuilder.run("mkdir -p " + containerHostingPath(emulatorConfig));
+                this.dockerBuilder.volume(containerHostingPath(emulatorConfig));
+            }
+        }
+
+        private void setupFunctions() {
+            if (emulatorConfig.firebaseConfig().functionsConfig().functionsPath.isPresent()) {
+                this.dockerBuilder.run("mkdir -p " + containerFunctionsPath(emulatorConfig));
+                this.dockerBuilder.volume(containerFunctionsPath(emulatorConfig));
+            }
+        }
+
+        private void setupUserAndGroup() {
```

Comment:
I'm not fan of creating user/group and chown, is it really necessary?

Reply:
Yes, that's what we discussed earlier in passing in the Github user/group, If you don't do this, docker will run with a different user (on CI builds at least, not for Docker Desktop), and you can;t access/read/remove/write the files from the host.